EIPS/eip-rich-site-proposed-contract-metadata.md

```diff
@@ -0,0 +1,93 @@
+1. Contract names
+  - Local address book
+    - Requires user interaction, which adds UX burden, and most users skip.
+  - Centralized registries
+    - Are perpetually incomplete, and favor the established, not the upstart.
+   - reverse-resolved ENS names
+     - Attackers can choose familiar-looking names, so these are not a reliable basis of secure review of what you're looking at unless you entered it yourself.
+2. Method information
+  - Registry lookups (4byte directory, the Parity on-chain registry)
+    - Can be prone to collisions (like At Inversebrah, with trolling increasing)
+    - Returns ABIs
+      - Has no parameter names, leaving most of the transaction illegible.
+    - Do not exist on all chains
+  - Trusting the contract for its own metadata (EIP-719, EIP-4430)
+    - Requires an active connection to the blockchain.
+    - Is not widespread (no backwards compatibility)
+  - Transaction Insights
+    - Requires an additional fairly heavy-weight process
+    - Is largely centralized today
+    - Only works for verified contracts on etherscan (although designed for eventual Sourcify support).
+
+Additionally, this method serves to help populate the wallet with trustworthy information about what it holds. It reduces the need to rely on any central services for indexing (enhancing performance and privacy), and eliminates dangers that come from index-based automatic asset detection, like receiving harrassment, or Airdrop scams.
+
+This approach can also work on cold/offline wallets.
+
+## Specification
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+Wallets may integrate the ability to propose names for contracts, methods, and parameters into any method that touches an address, so the wallet can be expanded with this metadata at runtime. These methods include `eth_sendTransaction`, and `eth_signTypedData` (and its variants).
+
+Any method that involves an address MAY include a new OPTIONAL parameter `proposedContracts`, which will include an additional prompt to the user as part of the confirmation flow, asking if they'd like to trust the site for this information.
+
+`contractId` would be some kind of cross chain address format.
+`ContractData` would be some kind of type that is enough to render transaction names for methods called.
+  - Today TX Insight uses a full Truffle Decoder output to render this, which includes lots of compiler info that we would not need here.
+```typescript
+type TxOptionsEnhancement = {
+  proposedContracts: {
+    [contractId: CrossChainAddress]: {
+      proposedName: string,
+      abi?: Abi,
+      contractData?: ContractData
+    }
+  }
+}
+```
+Wallets MUST store the recommending site & time along with this data for forensic purposes.
```

Comment:
```suggestion
  - Today TX Insight uses a full Truffle Decoder output to render this, which includes lots of compiler info that we would not need here.

```typescript
type TxOptionsEnhancement = {
  proposedContracts: {
    [contractId: CrossChainAddress]: {
      proposedName: string,
      abi?: Abi,
      contractData?: ContractData
    }
  }
}
e```

Wallets MUST store the recommending site & time along with this data for forensic purposes.
```
Please remove the `e` manually afterwards.